// datocms-client/lib/contentfulImport/command.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ora = _interopRequireDefault(require("ora"));

var _appClient = _interopRequireDefault(require("./appClient"));

var _getContentfulData = _interopRequireDefault(require("./getContentfulData"));

var _destroyExistingModels = _interopRequireDefault(require("./destroyExistingModels"));

var _destroyExistingAssets = _interopRequireDefault(require("./destroyExistingAssets"));

var _removeAllValidators = _interopRequireDefault(require("./removeAllValidators"));

var _setLocales = _interopRequireDefault(require("./setLocales"));

var _createModels = _interopRequireDefault(require("./createModels"));

var _createFields = _interopRequireDefault(require("./createFields"));

var _createRecords = _interopRequireDefault(require("./createRecords"));

var _addValidationsOnField = _interopRequireDefault(require("./addValidationsOnField"));

var _linkRecords = _interopRequireDefault(require("./linkRecords"));

var _createUploads = _interopRequireDefault(require("./createUploads"));

var _publishRecords = _interopRequireDefault(require("./publishRecords"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _default = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var contentfulToken, contentfulSpaceId, contentfulEnvironment, datoCmsToken, datoCmsEnvironment, datoCmsCmaBaseUrl, skipContent, contentType, client, datoClient, contentfulData, itemTypeMapping, fieldsMapping, _yield$createRecords, contentfulRecordMap, recordsToPublish, contentfulAssetsMap, linkedRecords, spinner;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            contentfulToken = _ref.contentfulToken, contentfulSpaceId = _ref.contentfulSpaceId, contentfulEnvironment = _ref.contentfulEnvironment, datoCmsToken = _ref.datoCmsToken, datoCmsEnvironment = _ref.datoCmsEnvironment, datoCmsCmaBaseUrl = _ref.datoCmsCmaBaseUrl, skipContent = _ref.skipContent, contentType = _ref.contentType;
            _context.prev = 1;
            _context.next = 4;
            return (0, _appClient["default"])(contentfulToken, contentfulSpaceId, datoCmsToken, datoCmsEnvironment, datoCmsCmaBaseUrl);

          case 4:
            client = _context.sent;
            datoClient = client.dato;
            _context.next = 8;
            return (0, _getContentfulData["default"])({
              client: client.contentful,
              skipContent: skipContent,
              contentType: contentType,
              contentfulEnvironment: contentfulEnvironment
            });

          case 8:
            contentfulData = _context.sent;
            _context.next = 11;
            return (0, _removeAllValidators["default"])({
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 11:
            _context.next = 13;
            return (0, _destroyExistingModels["default"])({
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 13:
            _context.next = 15;
            return (0, _destroyExistingAssets["default"])({
              datoClient: datoClient
            });

          case 15:
            _context.next = 17;
            return (0, _setLocales["default"])({
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 17:
            _context.next = 19;
            return (0, _createModels["default"])({
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 19:
            itemTypeMapping = _context.sent;
            _context.next = 22;
            return (0, _createFields["default"])({
              datoClient: datoClient,
              itemTypeMapping: itemTypeMapping,
              contentfulData: contentfulData
            });

          case 22:
            fieldsMapping = _context.sent;

            if (skipContent) {
              _context.next = 39;
              break;
            }

            _context.next = 26;
            return (0, _createRecords["default"])({
              itemTypeMapping: itemTypeMapping,
              fieldsMapping: fieldsMapping,
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 26:
            _yield$createRecords = _context.sent;
            contentfulRecordMap = _yield$createRecords.contentfulRecordMap;
            recordsToPublish = _yield$createRecords.recordsToPublish;
            _context.next = 31;
            return (0, _createUploads["default"])({
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 31:
            contentfulAssetsMap = _context.sent;
            _context.next = 34;
            return (0, _publishRecords["default"])({
              recordIds: recordsToPublish,
              datoClient: datoClient
            });

          case 34:
            _context.next = 36;
            return (0, _linkRecords["default"])({
              datoClient: datoClient,
              fieldsMapping: fieldsMapping,
              contentfulData: contentfulData,
              contentfulRecordMap: contentfulRecordMap,
              contentfulAssetsMap: contentfulAssetsMap
            });

          case 36:
            linkedRecords = _context.sent;
            _context.next = 39;
            return (0, _publishRecords["default"])({
              recordIds: linkedRecords,
              datoClient: datoClient
            });

          case 39:
            _context.next = 41;
            return (0, _addValidationsOnField["default"])({
              fieldsMapping: fieldsMapping,
              datoClient: datoClient,
              contentfulData: contentfulData
            });

          case 41:
            spinner = (0, _ora["default"])('Import completed! 🎉🎉🎉');
            spinner.succeed();
            _context.next = 49;
            break;

          case 45:
            _context.prev = 45;
            _context.t0 = _context["catch"](1);
            console.error('Importer error:', _context.t0);
            throw _context.t0;

          case 49:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[1, 45]]);
  }));

  return function (_x) {
    return _ref2.apply(this, arguments);
  };
}();

exports["default"] = _default;
